keps/sig-architecture/4330-compatibility-versions/README.md

```diff
@@ -507,6 +508,27 @@ compatibility version can be set to.
 
 Alpha APIs may not be enabled in conjunction with compatibility version.
 
+### Discovery
+
+Discovery will [enable](https://github.com/kubernetes/kubernetes/blob/7080b51ee92f67623757534f3462d8ae862ef6fe/staging/src/k8s.io/apiserver/pkg/util/openapi/enablement.go#L32) the group versions matching the compatibility version.
+
+API fields that were introduced after the compatibility version will **not** be
+pruned. There is a tradoff here:
+
+- If we keep the fields in the API, we leak information about fields that were
+  introduced in Kubernetes versions newer than the compatibility version.
+- If we remove the fields, we have an inconsistency between discovery and the
+  implementation, because the server implementation is aware of the field, and
+  may include it in responses (e.g. via defaulting).
+
+Note that even though we are deciding to include fields that were introduced
+after the compatibility version in discovery, new fields are always feature
+gated, so the fields will not settable and may not impact default behavior.
+
+Also note that we that show information about unavailable features in discovery
+today. We introduce fields into APIs for disabled-by-default features and make
+no attempt to hide those fields in discovery.
+
```

Comment:
This makes a lot of sense. The only extra thing I think is worth mentioning here that I'd be concerned about is what happens when an old version of API server deserializes an object at rest with unrecognized fields. I am not sure about our handling of that situation. I think if we an expect N+3 skip level upgrades in the future it may be reasonable to design for N-3 skip level downgrades to rollback such an update.

Reply:
skip level downgrades a whole can of worms.  I think round-trip-ability is the rub. Today we require all Beta and GA fields exist in the API for the N-1 version so that if we do a downgrade, we can still round trip data without data loss.

Taking the same approach for N-3 would imply that a feature can't be promoted to GA until it's been in alpha for 3 releases.. I don't want to ask the community to do that.

Reply:
> Taking the same approach for N-3 would imply that a feature can't be promoted to GA until it's been in alpha for 3 releases.. I don't want to ask the community to do that.

Hmm maybe you can correct my understanding, [elsewhere this KEP](https://github.com/kubernetes/enhancements/pull/4395/files#diff-92599d859941a447e4474b7bd83aa2ade15f76c12e7b0156ab24bcb09b1a4d87R205-R213) it says an N+3 upgrade is actually three N+1 upgrades using the N+3 binary.

1. Migrate to N+1 (using N+3 binary)
2. Migrate to N+2 (using N+3 binary)
3. Migrate to N+3  (using N+3 binary)

Wouldnt a downgrade be the same in reverse? Meaning we still would only need the N-1 release to know about the flag? If thats that case I think my original comment was asking if it's possible for the the N+3 binary to know which fields to drop when using `--compatibility=N+1` (if N+1 introduced the fields and N+2 was the first to use them).

Reply:
A good example might be the sidecars KEP.  A new field is introduced, and if that field is ever dropped, the pod becomes entirely invalid and won't do what it was suppose to do.  If you want to rollback a cluster using that feature, it's fundamentally unsafe to roll it back to a cluster version that is not aware of that field, because round tripping will drop the field and break the sidecar pods.

Reply:
Ah that makes sense, I think we had a great discussion here and offline about rollbacks that may benefit the KEP if summarized and put somewhere